arch/arc/core/cpu_idle.S

```diff
@@ -42,12 +42,22 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
 	POPR blink
 #endif
 
+/*
+ * Workaround for iotdk board. When iotdk board enter sleep mode,
+ * some peripherals (eg: uart) will power off and and board can't
+ * wake up through peripherals interrupt.
+ */
+#ifdef CONFIG_BOARD_IOTDK
```

Comment:
Is this specific to just the `BOARD_IOTDK` or the whole `SOC_ARC_IOT`?

Based on the comment you provided, this looks more like a SoC-level design flaw than a problem with this specific board. Is there a dedicated "wakeup" pin and is that pin not wired on that board or something?

Also in general, SoC (let alone board)-specific handling should not live in the arch code. If this is really necessary and going to be a common pattern, you should add an ARC-specific Kconfig (e.g. `ARC_DISABLE_SLEEP`) and select that from the relevant SoC or board.

Reply:
Right, so that currently suggested solution is a bit of an overkill :)

Basically we're dealing with peculiarities of a particular SoC where UART is a part of a power/clock domain which is gated on CPU sleep. That's a third-party design so I cannot tell if that decision to put UART in such a power/clock domain is good or not - I would say designers of that SoC knew better.

Thus I guess we need just to make sure we don't rely on UART IRQ for this SoC and only use ARC core Timers for waking-up the system.

Here we tried to cope with very slow serial terminal which was that slow due to the fact we were unable to get IRQs from UART. And while exclusion of `sleep` helps for obvious reasons (UART never gets gated), it's not something we should be doing here to mitigate gated UART during sleep.